test/integration_test/regression_test.go

```diff
@@ -0,0 +1,292 @@
+//go:build integrationtest
+// +build integrationtest
+
+package integrationtest
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	corev1 "github.com/libopenstorage/operator/pkg/apis/core/v1"
+	testutil "github.com/libopenstorage/operator/pkg/util/test"
+	"github.com/libopenstorage/operator/test/integration_test/types"
+	ci_utils "github.com/libopenstorage/operator/test/integration_test/utils"
+	appops "github.com/portworx/sched-ops/k8s/apps"
+	coreops "github.com/portworx/sched-ops/k8s/core"
+	"github.com/portworx/sched-ops/task"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var regressionTestCases = []types.TestCase{
+	{
+		TestName:        "NodeOutOfCpu",
+		TestrailCaseIDs: []string{""},
+		TestSpec: func(t *testing.T) interface{} {
+			// Construct StorageCluster
+			cluster := &corev1.StorageCluster{}
+			cluster.Name = clusterName
+			err := ci_utils.ConstructStorageCluster(cluster, ci_utils.PxSpecGenURL, ci_utils.PxSpecImages)
+			require.NoError(t, err)
+			return cluster
+		},
+		TestFunc: NodeOutOfCpu,
+	},
+}
+
+func shouldSkipRegressionTests(tc *types.TestCase) bool {
+	return false
+}
+
+func TestRegression(t *testing.T) {
+	for _, tc := range regressionTestCases {
+		tc.ShouldSkip = shouldSkipRegressionTests
+		tc.RunTest(t)
+	}
+}
+
+func NodeOutOfCpu(tc *types.TestCase) func(*testing.T) {
+	return func(t *testing.T) {
+		// Get the StorageCluster to start with
+		testSpec := tc.TestSpec(t)
+		cluster, ok := testSpec.(*corev1.StorageCluster)
+		require.True(t, ok)
+
+		// Pick a random node and label it with 'insufficient=cpu'
+		nodeNameWithLabel := ci_utils.AddLabelToRandomNode(t, "insufficient", "cpu")
+
+		// Create app deployment with node affinity to match above node
+		appDeployment, err := deployAppWithNodeAffinity()
+		require.NoError(t, err)
+
+		// Scale app 1 pod at a time until new pod is Pending due to '1 Insufficient cpu' message in status
+		err = scaleAppUntilInsufficientCpu(appDeployment)
+		require.NoError(t, err)
+
+		// Set resources for StorageCluster
+		resources := &v1.ResourceRequirements{
+			Requests: map[v1.ResourceName]resource.Quantity{
+				v1.ResourceMemory: resource.MustParse("2Gi"),
+				v1.ResourceCPU:    resource.MustParse("2000m"),
+			},
+		}
+		cluster.Spec.Resources = resources
+
+		// Deploy StorageCluster
+		logrus.Infof("Deploying StorageCluster once [%s]", cluster.Name)
+		cluster = ci_utils.DeployStorageCluster(cluster, ci_utils.PxSpecImages, t)
+
+		// Validate StorageCluster is failing
+		if err := testutil.ValidateStorageCluster(ci_utils.PxSpecImages, cluster, ci_utils.DefaultValidateDeployTimeout, ci_utils.DefaultValidateDeployRetryInterval, true); err == nil {
```

Comment:
do you think the timeout is too long here? if we are going to validate the pod out of cpu, probably we can skip this? and validate other pods are healthy underneath except one.

Reply:
Done